Mathlib/Analysis/InnerProductSpace/Adjoint.lean

```diff
@@ -393,6 +393,22 @@ theorem isStarProjection_iff_eq_starProjection_range [CompleteSpace E] {p : E 
   simpa [p.orthogonal_range, hp.isSelfAdjoint.isSymmetric]
     using congr($(hp.isIdempotentElem.mul_one_sub_self) x)
 
+theorem ContinuousLinearMap.IsStarProjection.mem_range_iff_norm {T : E →L[𝕜] E} [CompleteSpace E]
+    (hT : IsStarProjection T) {v : E} :
+    v ∈ LinearMap.range T ↔ ‖T v‖ = ‖v‖ := by
+  refine ⟨fun h => congr(‖$(LinearMap.IsIdempotentElem.mem_range_iff
+    congr(LinearMapClass.linearMap $hT.isIdempotentElem.eq) |>.mp h)‖), fun h => ?_⟩
+  have := calc 0 = ‖v‖ ^ 2 - ‖T v‖ ^ 2 := by simp [h]
+    _ = ‖T v + (1 - T) v‖ ^ 2 - ‖T v‖ ^ 2 := by simp
+    _ = ‖T v‖ ^ 2 + ‖(1 - T) v‖ ^ 2 - ‖T v‖ ^ 2 := by
+      congr
+      rw [norm_add_sq (𝕜 := 𝕜), ← adjoint_inner_right, hT.isSelfAdjoint.adjoint_eq]
+      simp [← mul_apply, hT.isIdempotentElem.eq]
+    _ = ‖(1 - T) v‖ ^ 2 := by simp
+  rw [eq_comm, sq_eq_zero_iff, norm_eq_zero, sub_apply, one_apply, sub_eq_zero, eq_comm] at this
+  exact LinearMap.IsIdempotentElem.mem_range_iff
+    congr(LinearMapClass.linearMap $hT.isIdempotentElem.eq) |>.mpr this
```

Comment:
We should have this variant of the previous theorem, so that when passing from left to right, we can substitute `T` with `K.starProjection` everywhere.
```lean
lemma isStarProjection_iff_eq_starProjection [CompleteSpace E] {p : E →L[𝕜] E} :
    IsStarProjection p ↔ ∃ K : Submodule 𝕜 E, ∃ (_ : K.HasOrthogonalProjection), p = K.starProjection :=
  ⟨fun h ↦ ⟨LinearMap.range p, isStarProjection_iff_eq_starProjection_range.mp h⟩,
    by rintro ⟨_, _, rfl⟩; simp⟩
```
Then you can do:
```suggestion
  obtain ⟨K, hK, rfl⟩ := isStarProjection_iff_eq_starProjection.mp hT
  simpa [h, sub_eq_zero, eq_comm (a := v), Submodule.starProjection_eq_self_iff] using
    K.norm_sq_eq_add_norm_sq_starProjection v
```